lib/pbench/test/unit/server/database/test_audit.py

```diff
@@ -0,0 +1,177 @@
+import datetime
+
+from freezegun.api import freeze_time
+import pytest
+from sqlalchemy.exc import IntegrityError
+
+from pbench.server import OperationCode
+from pbench.server.database.database import Database
+from pbench.server.database.models.audit import Audit, AuditNullKey, AuditStatus
+from pbench.server.database.models.datasets import Dataset
+from pbench.server.database.models.users import User
+from pbench.test.unit.server.database import FakeDBOrig, FakeRow, FakeSession
+
+
+class TestAudit:
+    session = None
+
+    def check_session(self, queries=0, committed: list[FakeRow] = [], rolledback=0):
+        """
+        Encapsulate the common checks we want to make after running test cases.
+
+        Args:
+            queries: A count of queries we expect to have been made
+            committed: A list of FakeRow objects we expect to have been
+                committed
+            rolledback: True if we expect rollback to have been called
+        """
+        session = self.session
+        assert session
+
+        # Check the number of queries we've created
+        assert len(session.queries) == queries
+
+        # 'added' is an internal "dirty" list between 'add' and 'commit' or
+        # 'rollback'. We test that 'commit' moves elements to the committed
+        # state and 'rollback' clears the list. We don't ever expect to see
+        # anything on this list.
+        assert not session.added
+
+        # Check that the 'committed' list (which stands in for the actual DB
+        # table) contains the expected rows.
+        assert sorted(list(session.committed.values())) == sorted(committed)
+
+        # Check whether we've rolled back a transaction due to failure.
+        assert session.rolledback == rolledback
+
+    @pytest.fixture(autouse=True, scope="function")
```

Comment:
Fixtures are such arcane black magic already I'm not sure it matters. I thought it was handy to patch the fake DB into all the test cases in the file automagically since that's what the thing is designed for. With great power comes great confusion...

Reply:
I'll grant that having foundational things laid in place automatically has the advantage that no one has to think about them...which is great until either someone wishes they weren't there or someone creates a new module which doesn't include the fixtures and wonders why they get different behavior.